rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,527 @@
+(3.3) can cause the migration to fail. The Kibana cluster should be able to
+recover automatically once these external conditions are resolved. There are
+two broad approaches to solving this problem based on whether or not
+migrations are idempotent: 
+
+| Idempotent migrations |Description                                                |
+| --------------------- | --------------------------------------------------------- |
+| Yes                   | Idempotent migrations performed without coordination      |
+| No                    | Single node migrations coordinated through a lease / lock |
+
+Idempotent migrations don't require coordination making the algorithm
+significantly less complex and will never require manual intervention to
+retry. We, therefore, prefer this solution, even though it introduces
+restrictions on migrations (4.3.1.1). For other alternatives that were
+considered see section [(5)](#5-alternatives).
+
+## 4.3.1 Idempotent migrations performed without coordination
+
+The migration system can be said to be idempotent if the same results are
+produced whether the migration was run once or multiple times. This property
+should hold even if new (up to date) writes occur in between migration runs
+which introduces the following restrictions:
+
+### 4.3.1.1 Restrictions
+
+1. All document transforms need to be deterministic, that is a document
+   transform will always return the same result for the same set of inputs. 
+2. It should always be possible to construct the exact set of inputs required
+   for (1) at any point during the migration process (before, during, after). 
+
+Although these restrictions require significant changes, it does not prevent
+known upcoming migrations such as [sharing saved-objects in multiple spaces](https://github.com/elastic/kibana/issues/27004) or [splitting a saved
+object into multiple child
+documents](https://github.com/elastic/kibana/issues/26602). To ensure that
+these migrations are idempotent, they will have to generate new saved object
+id's deterministically with e.g. UUIDv5.
+
+### 4.3.1.2 Algorithm
+
+1. On startup, find the index to which the `.kibana` alias points,
+   `.kibana_n`, and check if any documents have an outdated `migrationVersion`
+   indicating that a migration needs to be performed. 
+2. Create `.kibana_n+1`, if it already exists, ignore the error and continue.
+3. Migrate documents by reading from `.kibana_n` and creating documents in
+   `.kibana_n+1`. If a document already exists, ignore.
+4. Move the `.kibana` index to point to `.kibana_n+1` (ignore if it has
+   already been moved). 
+
+Together with the limitations this algorithm ensures that migrations are
+idempotent. If two nodes are started simultaneously, both of them will start
+writing into `.kibana_n+1` but because migrations are idempotent, it doesn’t
+matter which node’s writes win.
+
+> Note: This algorithm doesn't support mixed Kibana versions attempting to
```

Comment:
When the index is first created, we also specify the mappings. If the migration was started by version 7.9, and resumed by version 7.10, the mappings are likely to change and the migration will potentially fail mid-migration.

To mitigate against this mid-migration failure and allow version 7.9 to resume the migration, what if we include the version number in the mappings metadata? When the migrations are on step 2 and we're unable to create the index, we can retrieve the mappings and ensure we're resuming the migrations for the same version. This potentially limits the situations in which we can resume an abandoned or failed migration, while preventing against us resulting in an inconsistent state.

This same logic **could** be applied when we have different plugins. I'm honestly not sure what the result is of multiple instances with different plugins both running the migrations to know whether or not this warrants the same protections.

Reply:
Yes I think we need to add mappings verification step if step 2 fails to create an index. The mappings metadata contains a hash of all the mappings per saved object type so we already have enough information to know whether the mappings are compatible or not. I agree with you, if they're not compatible we should just fail the migration because another node could be busy migrating and one node failing hard is better than an inconsistent state.

To calculate this "compatibility" the existing system already considers mappings that are defined on the index but not in the known mappings on the node as "compatible". This means if a node with a plugin enabled starts a migration and then later a node with that plugin disabled comes online, it will successfully continue the migration. However the opposite isn't true. If a plugin was disabled when the migration started and then enabled, the later node will consider the mappings as being incompatible.

The mappings of different enabled plugins should always be additive so if we could get the two nodes to synchronize what the union of their mappings should be we could allow the migration to proceed in parallel. But this would require something like optimistic concurrency control on the mappings API to avoid race conditions. Maybe there is a brute force approach where we update the mappings hoping the other node doesn't override our additions and then start migrating objects, if a write fails due to a mappings error we start over, updating the mappings again and again trying write our migrated objects.

Reply:
This mapping synchronization is also a problem in 8.x when there are mixed versions performing a migration in parallel or nodes on the same version with different enabled plugins.

I don't think this is a very common occurrence, so it feels like failing hard and instructing the user to recover a previous snapshot in order to retry would be acceptable.

Reply:
I've added a mappings compatibility check.